src/Network/AWS/ARN/S3.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_HADDOCK show-extensions #-}
+
+-- |
+--
+-- Module      : Network.AWS.ARN.S3
+-- Copyright   : (C) 2020-2023 Bellroy Pty Ltd
+-- License     : BSD-3-Clause
+-- Maintainer  : Bellroy Tech Team <haskell@bellroy.com>
+-- Stability   : experimental
+module Network.AWS.ARN.S3
+  ( -- * S3 Object
+    Object (..),
+    parseObject,
+    renderObject,
+
+    -- ** Prisms
+    _Object,
+
+    -- ** Conversions to bucket names
+    objectInBucket,
+    containingBucket,
+
+    -- * S3 Bucket
+    Bucket (..),
+    parseBucket,
+    renderBucket,
+
+    -- ** Prisms
+    _Bucket,
+  )
+where
+
+import Data.Hashable (Hashable)
+import Data.Text (Text)
+import qualified Data.Text as T
+import GHC.Generics (Generic)
+import Network.AWS.ARN.Internal.Lens (Prism', prism')
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+-- >>> import Network.AWS.ARN.Internal.Lens ((^?))
+
+-- | An AWS S3 object, made of a bucket and an object key.
+--
+-- >>> "bucket-name/my/object" ^? _Object
+-- Just (Object {bucketName = "bucket-name", objectKey = "my/object"})
+
+-- >>> "bucket-name" ^? _Object
+-- Nothing
+
+-- >>> containingBucket $ "bucket-name/my/object" ^? _Object
+-- Just (Bucket {bucketName = "bucket-name"})
+-- @since 0.3.1.0
+data Object = Object
+  { bucketName :: Text,
+    objectKey :: Text
+  }
+  deriving (Eq, Ord, Hashable, Show, Generic)
+
+-- @since 0.3.1.0
+objectInBucket :: Text -> Bucket -> Object
+objectInBucket objectKey (Bucket bucketName) = Object bucketName objectKey
+
+-- @since 0.3.1.0
+containingBucket :: Object -> Bucket
```

Comment:
This could be a `Lens' Object Bucket`?

Reply:
In fact, if you make Object contain a Bucket instead of a bucket name, you get `#bucket` from the Generic instance. Not sure if you would still want a `bucketName :: Lens' Object Text`?

Reply:
`Object` containing `Bucket` sounds like a good idea.

Reply:
This function is redundant; `bucket` is already exposed in the module header.